### ISSUE TITLE
Add namespace support to ParseRule

### ISSUE THREAD

**Issue by vandenoever** [ProseMirror/prosemirror#643]
I request that ParseRule is extended to support matching by QName. Webpages can use many namespaces as in RDF, SVG etc.

ParseRule can currently use CSS selectors via `tag` and inline styles via `style`. This not enough to take XML namespaces into account.

While CSS has support for selection by namespaces via `@namespace`, this does not work in the `matches()` function of the DOM. The namespace declaration is only active in the stylesheet in which it is declared. So an expression like `xlink|a` gives an error even when the document has defined `xmlns:xlink`.

I propose to add an additional field to ParseRule named either `ns` or `namespace` which matches against the namespaceURI of the element and can be used with the `tag` property.

This would require an extra clause in the if statement in `matchTag(dom, context) `.

https://github.com/ProseMirror/prosemirror-model/blob/master/src/from_dom.js#L167

**Comment by marijnh** [comment]
Does the DOM even make it possible to match selectors with namespaces? I wouldn't know how.

**Comment by vandenoever** [comment]
In CSS stylesheets namespaces work great. That's core part of how WebODF works. But in the DOM, they do not [1] That's why I propose to add a namespace property to ParseRule.

[1] https://dom.spec.whatwg.org/#selectors "Support for namespaces within selectors is not planned and will not be added."

**Comment by marijnh** [comment]
I don't see how you expect this property to be implemented. We're using the DOM's selector matching code, which does not support namespaces.

**Comment by vandenoever** [comment]
Here's a screenshot that show DOM elements with namespaces that are styled via CSS with namespace support.

![css_namespaces](https://cloud.githubusercontent.com/assets/608417/25909787/8250ea3e-35ae-11e7-99d0-e3057aafd3bb.png)

Matching namespaces in addition to CSS selectors can be done like this:

```js
if (matches(dom, rule.tag) && matchesNamespace(dom, rule.namespace) && (!rule.context || context.matchesContext(rule.context))) {
```

```js
matchesNamespace(dom, namespace) {
  return dom.namespaceURI == namespace
}
```

Matching on multiple namespaces can be done by matching with multiple ParseRules.

**Comment by marijnh** [comment]
So a DOM node's `match` method will completely ignore the namespace?

**Comment by vandenoever** [comment]
The `matches` method seems to ignore it and gives an error message if you try to use one.

E.g. in the console

```js
document.documentElement.matches("*|html")
true
document.documentElement.matches("html|html")
SyntaxError: 'html|html' is not a valid selector
```

**Comment by vandenoever** [comment]
Note: CSS namespaces uses '|' instead of ':' so `svg:g` becomes `svg|g`.

**Comment by vandenoever** [comment]
"E if no default namespace has been declared for selectors, this is equivalent to *|E."

https://drafts.csswg.org/selectors-4/#type-nmsp

**Comment by marijnh** [comment]
That seems to indeed work. Odd, but useful in this case. Do you want to submit a pull request implementing this feature? I prefer the property name `namespace`.

**Comment by vandenoever** [comment]
Sure.

**Comment by vandenoever** [comment]
There she blows: https://github.com/ProseMirror/prosemirror-model/pull/9

**Comment by marijnh** [comment]
Merged ProseMirror/prosemirror-model#9
